extends Area2D

var travelled_distance = 0

const SPEED = 500
const RANGE = 1600
@export var damage := 1.0
const EXPLOSION = preload("res://effects/explosion_effect.tscn")


func _physics_process(delta: float) -> void:
	var direction = Vector2.RIGHT.rotated(rotation)
	position += direction * SPEED * delta
	
	travelled_distance += SPEED * delta
	if travelled_distance > RANGE:
		queue_free()


func _on_body_entered(body: Node2D) -> void:
	# Spawn explosion TR∆Ø·ªöC KHI x√≥a ƒë·∫°n
	queue_free()
	if body.has_method("take_dame"):
		body.take_dame(damage)
		spawn_explosion()



func spawn_explosion() -> void:
	# T·∫°o explosion
	var explosion = EXPLOSION.instantiate()
	
	# ƒê·∫∑t v·ªã tr√≠ = v·ªã tr√≠ ƒë·∫°n
	explosion.global_position = global_position
	
	# C√≥ th·ªÉ random xoay cho ƒë·∫πp
	explosion.rotation = randf() * TAU
	
	# M√†u cho n·ªï l·ª≠a
	#explosion.modulate = Color(0.759, 0.322, 0.475, 1.0)
	
	# Random scale
	var random_scale = randf_range(3, 4)
	explosion.scale = Vector2(random_scale, random_scale)
	
	# Add v√†o scene ch√≠nh (kh√¥ng ph·∫£i v√†o bullet!)
	get_tree().current_scene.add_child(explosion)
	
	print("üí• Explosion created at: ", global_position)
